### PR TITLE
Release Charm Action

### PR DESCRIPTION
**Pull request by agathanatasha**
Release action releases a revision or a revision given the `track/risk` channel name to a specified destination channel.
If both revision and origin channel are filled in, revision number takes precedent.
It finds the revision number of a channel by parsing the output of `charmcraft status`.
It parses the combination of resources versions from github releases. Therefore, the charm must use `upload-charm` action to publish.
After releasing to charmcraft, it updates the github release of that revision with the destination channel and a timestamp. The line would be added at the end of the release body before the github generated changelog.
e.g. `Released to 'latest/candidate' at 21:10 UTC on 5 Apr 2022`

Limitations:
This action doesn't work with charm that support multiple bases.
The origin channel must be in the format of `track/risk` for parsing the charmcraft status output.
This action is only individual charms. It does not support releasing bundles.
It does not support charm with channel branches as it would mess up the charmcraft status output.

### CODE REVIEW

File: release-charm/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        required: false
+      rev:
+        description: 'Revision number'
+        required: false
+      tag-prefix:
+        description: 'Tag Prefix'
+        required: false
+      charm-path:
+        description: 'Path to charm'
+        required: false
```
Comment (simskij) [inline release-charm/README.md:27]: Why are these needed? We should only promote what is already published in some other channel. To me, it makes sense to keep the workflow strict and not allow too much deviation or customizability.

**Review by simskij — CHANGES_REQUESTED** [review]
Wow, this is awesome! Thank you! I do, however, have some comments that I'd like us to discuss prior to accepting it.

File: release-charm/README.md (hunk 1 above)
Comment (agathanatasha) [reply release-charm/README.md:27]: This action was designed to be triggered manually with certain input. I think it would be helpful to include the yaml for inputs as reference. 
The key inputs are destination channel and origin channel/revision. I don't know a way to require either origin channel or revision. That's why I left both as not required.
tag-prefix and charm-path are for repos with multiple charm. You could omit those fields for single charm repos.

File: src/actions/release-charm/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  const result = resourceNames.reduce((acc, field, index) => {
+    acc.push({ resourceName: field, resourceRev: resourceRevs[index] });
+    return acc;
+  }, [] as Array<{ resourceName: string; resourceRev: string }>);
+  return result;
+}
+
+export { getResourcesInfoByRelease };
```
[the remark this answers is not in the document]
Comment (agathanatasha) [reply src/actions/release-charm/helpers.ts:35]: Since this action accepts either origin channel or revision number, release body is the only place with all the information. If I want to release rev5 to latest/beta, but rev5 is not the latest revision on any channels, I won't be able to find the combination of resources used.

File: release-charm/README.md (hunk 1 above)
Comment (ca-scribner) [reply release-charm/README.md]: I had some similar comments over [here](https://github.com/canonical/istio-operators/pull/80/files/11b49e2ffa3145b2658d461222fb2214f3f991c8#r844119799).  Summary: if rev number and origin channel are mutually exclusive, we should highlight that in their descriptions and the actual code should error out if it receives both (otherwise people will accidentally specify both and not know what result they'll receive).

File: src/actions/release-charm/helpers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ca-scribner) [reply src/actions/release-charm/helpers.ts:35]: @agathanatasha, I think you're saying that we may need to promote something that is not currently in any channel (eg: promote a new revision to edge), `charmcraft status` only gives details of charms currently in a channel?

File: release-charm/README.md (hunk 1 above)
Comment (ca-scribner) [reply release-charm/README.md]: But more fundamentally, @simskij I see you're wondering why we need `revision number` at all.  I think this is needed for the case where something is not yet in a channel at all.

File: src/actions/release-charm/helpers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (agathanatasha) [reply src/actions/release-charm/helpers.ts:35]: That's correct. `charmcraft status` only show resources details for the latest revision in a channel.

[the remark this answers is not in the document]
Comment (ca-scribner) [reply src/actions/release-charm/helpers.ts:35]: between this and the above comment, I think the discussion is really about scope.  Should this action handle cases of publishing to a channel, or just promoting from one channel to another.

[the remark this answers is not in the document]
Comment (DomFleischmann) [reply src/actions/release-charm/helpers.ts:35]: I believe that us having the flexibility of releasing any revision to a channel is going to be very important. For one it would allow us to rollback a release that had a breaking change. Additionally, we don't have to worry about our channels changing between the time we decide promoting something to a stable channel and us clicking the button.

File: release-charm/README.md (hunk 1 above)
Comment (simskij) [reply release-charm/README.md]: @ca-scribner @agathanatasha I don't think we should allow publishing something not already in a channel. upload-charm will put it on edge, and then that revision should be what gets promoted. Opening up for skipping over channels is a disaster waiting to happen.

File: src/actions/release-charm/helpers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (simskij) [reply src/actions/release-charm/helpers.ts:35]: I really want to stress that I think @agathanatasha has done some great work here and demonstrated a lot of ingenuity in solving this.

> (eg: promote a new revision to edge)

That is already done automatically by the upload-charm action.

> I believe that us having the flexibility of releasing any revision to a channel is going to be very important. 

Parsing the content of a tag message that may or may not be there and using that to make decisions in our release automation just isn't within the realm of what I consider an acceptable risk. The value of that flexibility just isn't anywhere near outweighing the added fragility. 

> it would allow us to rollback a release that had a breaking change.

Promoting something to `stable` and then having to do a rollback is something I imagine happening very seldom, if ever. For all other channels, fixing forward is a perfectly acceptable workflow. In my opinion, the frequency of this practice does not warrant taking it into consideration.

> Additionally, we don't have to worry about our channels changing between the time we decide to promote something to a stable channel and us clicking the button.

If the risk for the content of a channel changing between you deciding to promote something and clicking the button is real, I don't think that will be solved by technology.

File: release-charm/README.md (hunk 1 above)
Comment (ca-scribner) [reply release-charm/README.md]: @simskij I disagree.  Can you give an example of the disaster you're thinking of?  I'm having trouble thinking of a scenario where putting a charm from unpublished to `candidate`, etc., has a bad consequence

The example that comes to mind for why this is needed is:
1. a new revision of the charm is created and put into charmhub (`rev1`) on `latest/edge`
2. we start doing QA to see if `rev1` is good for release (ask field, push to another team member, etc)
3. while doing QA, someone finishes a new feature not intended for a near-term release and pushes it to main.  This triggers CI to push `rev2` to `latest/edge`
4. QA process passes and we want to release `rev1` to `latest/stable`

How can we complete the release in step (4)?

Comment (ca-scribner) [reply release-charm/README.md]: A bit removed from this conversation, but I find it off that are dividing our release CI building blocks into:
* upload-charm: build, upload to charmhub, release to edge
* release: release to track/[^edge]

By convention it totally makes sense that our publish CI build, upload, and release to edge because we decided that `latest/edge` should track with `main`.  But those are two basic actions, not one.  I don't think that means releasing to edge should be tightly coupled with build/upload.  I think the name of the action alone is a good guide - `upload-charm` does not suggest anything is being released, and I can't think of another natural name that doesn't have "and" in it (eg: `upload-and-release-charm`)

This makes more sense to me:

upload-charm: build, upload to charmhub
release-charm: release to anywhere (either from unreleased to a channel, or from one channel to another)

I like that our publish CI uploads and releases to `latest/edge` so I propose we keep that, just that it shouldn't be one action.  publish CI could just do both

File: src/actions/release-charm/helpers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (DomFleischmann) [reply src/actions/release-charm/helpers.ts:35]: > Parsing the content of a tag message that may or may not be there and using that to make decisions in our release automation just isn't within the realm of what I consider an acceptable risk. The value of that flexibility just isn't anywhere near outweighing the added fragility.

I completely agree that parsing the tag messages is not a good way of handling this. But us currently being limited in the way we can talk to charmhub and coming up with this suboptimal workaround  does not mean the initial use case is invalid. I still believe that being able to promote certain revisions is an important use case, specially considering how Field relies heavily on pinning revisions for any kind of customer deployments. 

To move this PR forward we can either leave it like it is or remove the revision specific promotion part and wait until charmcraft allows us to get more details from a certain charm revision even if it is not currently in a channel. @jnsgruk what are your thought on this?

**Comment by simskij** [comment]
We had a call today to try to iron out how to proceed with this one. The conclusion we've arrived at is:

1) We're only gonna allow promoting revisions **already in a channel**.
2) We need to switch the resource revision fetch from using the git tags to using `charmcraft status`

Thanks again for all the input!

File: release-charm/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+| `tag-prefix`         | Tag prefix, useful when bundling multiple charms in the same repo using a matrix.                       |          |     
+| `charm-path`         | Path to the charm where `metadata.yaml` is located. Defaults to the current working directory.      |          |    
+| `charmcraft-channel` | Snap channel to use when installing charmcraft. Defaults to `latest/edge`.                              |          |
+
+### Outputs
+
+None
+
+### Limitations
+- Doesn't work with charm that support multiple bases
```
Comment (simskij) [inline release-charm/README.md:55]: What does this mean?

Comment (agathanatasha) [reply release-charm/README.md:55]: The base column in charmcraft status, e.g. `ubuntu 20.04 (amd64)`. If a charm support multiple bases, there would be one latest track for each base.
https://pastebin.ubuntu.com/p/gDmXc2b7fK/

File: release-charm/README.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
```
Comment (simskij) [inline release-charm/README.md:63]:
```suggestion
- Doesn't work with charm that are built on multiple bases
```

**Review by simskij — APPROVED** [review]
